### PR TITLE
[MTDSA-10969]Updated SubmitSelfEmploymentParser

### PR DESCRIPTION
**Pull request by vfryer12**

### CODE REVIEW

File: test/v3/controllers/requestParsers/validators/SubmitSelfEmploymentBsasValidatorSpec.scala [hunk 1]
```diff
@@ -57,117 +61,127 @@ class SubmitSelfEmploymentBsasValidatorSpec extends UnitSpec {
...
+          calculationId,
+          AnyContentAsJson(Json.obj("income" -> "{}")))
+        ) shouldBe List(RuleIncorrectOrEmptyBodyError.copy(paths = Some(Seq("/income"))))
+      }
+
+      "fields are empty" in {
+        validator.validate(SubmitSelfEmploymentBsasRawData(
+          nino,
+          calculationId,
+          AnyContentAsJson(Json.toJson(defaultBody(income = Some(Income(None, None))))))
```
Comment (jeremystone) [inline test/v3/controllers/requestParsers/validators/SubmitSelfEmploymentBsasValidatorSpec.scala:127]: Can't really rely on validating based on model objects converted to JSON since the Writes are used for downstream requests. You would be much safer constructing the necessary small chunk of JSON in-line here.

File: test/v3/controllers/requestParsers/validators/SubmitSelfEmploymentBsasValidatorSpec.scala [hunk 2]
```diff
@@ -16,19 +16,23 @@
...
 class SubmitSelfEmploymentBsasValidatorSpec extends UnitSpec {
   val validator = new SubmitSelfEmploymentBsasValidator()
 
-  val bsasId   = "a54ba782-5ef4-47f4-ab72-495406665ca9"
-  val nino     = "AA123456A"
-  val seIncome: Income = Income(Some(100.49), Some(100.49))
+  val calculationId = "a54ba782-5ef4-47f4-ab72-495406665ca9"
+  val nino = "AA123456A"
+
+  val seIncome: Income = Income(
```
Comment (jeremystone) [inline test/v3/controllers/requestParsers/validators/SubmitSelfEmploymentBsasValidatorSpec.scala:32]: Remove seIncome, seExpenses, seAdditions and defaultBody - see above comment.

File: test/v3/controllers/requestParsers/validators/SubmitSelfEmploymentBsasValidatorSpec.scala [hunk 3]
```diff
@@ -57,117 +61,127 @@ class SubmitSelfEmploymentBsasValidatorSpec extends UnitSpec {
...
-        validator.validate(SubmitSelfEmploymentBsasRawData(nino, bsasId,
-          AnyContentAsJson(Json.toJson(defaultBody(income = Some(Income(None, None))))))) shouldBe List(RuleIncorrectOrEmptyBodyError)
-        validator.validate(SubmitSelfEmploymentBsasRawData(nino, bsasId,
-          AnyContentAsJson(Json.obj()))) shouldBe List(RuleIncorrectOrEmptyBodyError)
-        validator.validate(SubmitSelfEmploymentBsasRawData(nino, bsasId,
-          AnyContentAsJson(mtdRequestWithBothExpenses))) shouldBe List(RuleBothExpensesError)
-        validator.validate(SubmitSelfEmploymentBsasRawData(nino, bsasId,
-          AnyContentAsJson(mtdRequestWithAdditionsAndExpenses))) shouldBe List(RuleBothExpensesError)
+
+      "object is empty" in {
```
Comment (jeremystone) [inline test/v3/controllers/requestParsers/validators/SubmitSelfEmploymentBsasValidatorSpec.scala:115]: Need separate tests for `expenses` and `additions` fields being empty too.

File: test/v3/controllers/requestParsers/validators/SubmitSelfEmploymentBsasValidatorSpec.scala [hunk 4]
```diff
@@ -57,117 +61,127 @@ class SubmitSelfEmploymentBsasValidatorSpec extends UnitSpec {
...
+        validator.validate(SubmitSelfEmploymentBsasRawData(
+          nino,
+          calculationId,
+          AnyContentAsJson(mtdRequestWithBothExpenses))
+        ) shouldBe List(RuleBothExpensesError)
       }
+    }
+
+    "return ValueFormatError" when {
+      "multiple fields are invalid" in {
```
Comment (jeremystone) [inline test/v3/controllers/requestParsers/validators/SubmitSelfEmploymentBsasValidatorSpec.scala:151]: Need tests for single fields being invalid too. 

(See `SubmitForeignPropertyBsasValidatorSpec` for one way to do this relatively compactly. Basically have some JSON with everything in it and update each field in turn to bad values using methods in the `JsonErrorValidators` trait.)

File: test/v3/controllers/requestParsers/validators/SubmitSelfEmploymentBsasValidatorSpec.scala [hunk 5]
```diff
@@ -57,117 +61,127 @@ class SubmitSelfEmploymentBsasValidatorSpec extends UnitSpec {
...
-        validator.validate(SubmitSelfEmploymentBsasRawData(nino, bsasId,
-          AnyContentAsJson(mtdRequestWithBothExpenses))) shouldBe List(RuleBothExpensesError)
-        validator.validate(SubmitSelfEmploymentBsasRawData(nino, bsasId,
-          AnyContentAsJson(mtdRequestWithAdditionsAndExpenses))) shouldBe List(RuleBothExpensesError)
+
+      "object is empty" in {
+        validator.validate(SubmitSelfEmploymentBsasRawData(
+          nino,
+          calculationId,
+          AnyContentAsJson(Json.obj("income" -> "{}")))
```
Comment (jeremystone) [inline test/v3/controllers/requestParsers/validators/SubmitSelfEmploymentBsasValidatorSpec.scala:119]: This results in JSON:
``` json
{
  "income": "{}"
}
```
i.e. income is a string value not an empty object.  

Suspect it should be:
```suggestion
          AnyContentAsJson(Json.obj("income" -> JsObject.empty)))
```

File: app/v3/controllers/requestParsers/validators/SubmitSelfEmploymentBsasValidator.scala [hunk 6]
```diff
@@ -52,78 +57,174 @@ class SubmitSelfEmploymentBsasValidator extends Validator[SubmitSelfEmploymentBs
...
+
+  private def bothExpensesValidation: SubmitSelfEmploymentBsasRawData => List[List[MtdError]] = { data =>
     val model: SubmitSelfEmploymentBsasRequestBody = data.body.json.as[SubmitSelfEmploymentBsasRequestBody]
 
     List(
       if(model.expenses.exists(_.isBothSupplied)
-        || (!(model.additions.isEmpty || model.additions.exists(_.isEmpty))
-        && (!(model.expenses.isEmpty || model.expenses.exists(_.isConsolidatedExpensesEmpty))))){
+        || (!(model.expenses.isEmpty || model.expenses.exists(_.isConsolidatedExpensesEmpty))
+        && (!(model.additions.isEmpty || model.additions.exists(_.isEmpty))))){
```
Comment (jeremystone) [inline app/v3/controllers/requestParsers/validators/SubmitSelfEmploymentBsasValidator.scala:227]: It is not obvious that this is correct and there do not appear to be test cases that exercise the second two lines (comment them out and the tests still pass).

I would write some extra tests first (maybe move the code to the existing BothExpensesValidation/BothExpensesValidationSpec where testing exhaustively will prob be easier) but you might be able to simplify the logic to something like:
```scala
      if (model.expenses.exists(_.isBothSupplied)
          || (
            model.expenses.exists(x => x.consolidatedExpenses.isDefined) &&
            model.additions.exists(x => !x.isEmpty)
          )) { ...
```

**Review by jeremystone — CHANGES_REQUESTED** [review]
A few (overwhelmingly pre-existing 😬) problems to fix.

File: app/v3/models/request/submitBsas/selfEmployment/SubmitSelfEmploymentBsasRequestBody.scala [hunk 7]
```diff
@@ -19,13 +19,11 @@ package v3.models.request.submitBsas.selfEmployment
...
+case class SubmitSelfEmploymentBsasRequestBody(income: Option[Income],
+                                               expenses: Option[Expenses],
+                                               additions: Option[Additions]){
 
-  def isEmpty: Boolean = !(income.isDefined || additions.isDefined || expenses.isDefined)
-
-  def isIncorrectOrEmptyBodyError: Boolean = isEmpty || (income.isDefined && income.get.isEmpty) ||
-    (additions.isDefined && additions.get.isEmpty) ||
-    (expenses.isDefined && expenses.get.isEmpty)
+  def isEmpty: Boolean = !(income.isDefined || expenses.isDefined || additions.isDefined)
```
Comment (mattstephens-hmrc) [inline app/v3/models/request/submitBsas/selfEmployment/SubmitSelfEmploymentBsasRequestBody.scala:26]: I think this line was already there, but could it be:

`val isEmpty: Boolean = income.isEmpty && expenses.isEmpty && additions.isEmpty`

**Review by jeremystone — APPROVED** [review]
